File: 39. Combination Sum/39. Combination Sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    let res = []

    var backtrack = function(start, path, total){
        if (total == target){
            res.push([...path])
            return
        }
        if (total > target) return

        for (let i = start; i < candidates.length; i++){
            path.push(candidates[i])
            backtrack(i, path, total + candidates[i])
            path.pop()
        }
    }

    backtrack(0, [], 0)
    return res
};

File: 39. Combination Sum/39. Combination Sum.c
#include <stdlib.h>
#include <string.h>

// Helper function
void backtrack(int* candidates, int candidatesSize, int target,
               int* path, int pathSize,
               int*** result, int* returnSize, int** returnColumnSizes,
               int start) {
    
    if (target == 0) {
        // Allocate memory for a new combination
        int* combination = (int*)malloc(sizeof(int) * pathSize);
        memcpy(combination, path, sizeof(int) * pathSize);

        // Store this combination
        (*result)[*returnSize] = combination;
        (*returnColumnSizes)[*returnSize] = pathSize;
        (*returnSize)++;
        return;
    }

    for (int i = start; i < candidatesSize; i++) {
        if (candidates[i] > target) continue;

        path[pathSize] = candidates[i];
        backtrack(candidates, candidatesSize, target - candidates[i],
                  path, pathSize + 1, result, returnSize, returnColumnSizes, i);
    }
}

int** combinationSum(int* candidates, int candidatesSize, int target, int* returnSize, int** returnColumnSizes) {
    int** result = (int**)malloc(sizeof(int*) * 200);             // Allocate space for up to 200 combinations
    *returnColumnSizes = (int*)malloc(sizeof(int) * 200);         // Column sizes for each result row
    *returnSize = 0;

    int* path = (int*)malloc(sizeof(int) * 100);                  // Temporary array to hold current combination

    backtrack(candidates, candidatesSize, target, path, 0, &result, returnSize, returnColumnSizes, 0);

    free(path); // clean up temporary path
    return result;
}
